src/hotspot/share/classfile/stringTable.cpp

```diff
@@ -793,7 +793,11 @@ void StringTable::allocate_shared_strings_array(TRAPS) {
     size_t secondary_array_size = objArrayOopDesc::object_size(_secondary_array_max_length);
 
     if (ArchiveHeapWriter::is_too_large_to_archive(secondary_array_size)) {
-      fatal("Too many strings to be archived: " SIZE_FORMAT, _items_count);
+      // This can only happen if you have an extremely large number of classes that
+      // refer to more than 16384 * 16384 = 26M interned strings! Not a practical concern
+      // but bail out for safety.
+      log_error(cds)("Too many strings to be archived: " SIZE_FORMAT, _items_count);
+      os::_exit(1);
```

Comment:
Hmm I hadn't really noticed that the CDS code has introduced its own direct exit path using `os::_exit` in a number of places. Why does it do this instead of using the existing VM exit paths?

Reply:
The reason is explained in metaspaceShared.cpp, but perhaps we should consolidate all these direct exit calls (in a separate PR) to an utility function so it's easy to track (and explain) all the calls.

```
  // There may be pending VM operations. We have changed some global states
  // (such as vmClasses::_klasses) that may cause these VM operations
  // to fail. For safety, forget these operations and exit the VM directly.
  os::_exit(0);
```

Reply:
`vm_direct_exit` would seem to be okay in that case.